samza-core/src/main/scala/org/apache/samza/diagnostics/DiagnosticsManager.java

```diff
@@ -0,0 +1,173 @@
+  private final String samzaVersion;
+  private final String hostname;
+  private final Instant resetTime;
+
+  private SystemProducer systemProducer; // SystemProducer for writing diagnostics data
+  private BoundedList<DiagnosticsExceptionEvent> exceptions; // A BoundedList for storing DiagnosticExceptionEvent
+  private final ScheduledExecutorService scheduler; // Scheduler for pushing data to the diagnostic stream
+  private final SystemStream diagnosticSystemStream;
+
+  public DiagnosticsManager(String jobName, String jobId, String containerName, String executionEnvContainerId,
+      String taskClassVersion, String samzaVersion, String hostname, SystemStream diagnosticSystemStream,
+      SystemProducer systemProducer) {
+    this.jobName = jobName;
+    this.jobId = jobId;
+    this.containerName = containerName;
+    this.executionEnvContainerId = executionEnvContainerId;
+    this.taskClassVersion = taskClassVersion;
+    this.samzaVersion = samzaVersion;
+    this.hostname = hostname;
+    resetTime = Instant.now();
+
+    this.systemProducer = systemProducer;
+    this.diagnosticSystemStream = diagnosticSystemStream;
+    this.exceptions = new BoundedList<>("exceptions"); // Create a BoundedList with default size and time parameters
+    this.scheduler = Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder().setNameFormat(PUBLISH_THREAD_NAME).setDaemon(true).build());
+
+    try {
+
+      Util.getObj("org.apache.samza.logging.log4j.SimpleDiagnosticsAppender",
+          JavaConverters.collectionAsScalaIterableConverter(
+              Collections.singletonList(new Tuple2<Class<?>, Object>(DiagnosticsManager.class, this)))
+              .asScala()
+              .toSeq());
+
+      LOG.info("Attached log4j diagnostics appender.");
+    } catch (ClassNotFoundException | InstantiationException | InvocationTargetException e) {
+      try {
+        Util.getObj("org.apache.samza.logging.log4j2.SimpleDiagnosticsAppender",
+            JavaConverters.collectionAsScalaIterableConverter(
+                Collections.singletonList(new Tuple2<Class<?>, Object>(DiagnosticsManager.class, this)))
+                .asScala()
+                .toSeq());
+        LOG.info("Attached log4j diagnostics appender.");
+      } catch (ClassNotFoundException | InstantiationException | InvocationTargetException ex) {
+
+        LOG.warn(
+            "Failed to instantiate neither diagnostic appender for sending error information to diagnostics stream.",
+            ex);
+      }
+    }
+  }
+
+  public void start() {
+    this.scheduler.scheduleWithFixedDelay(new DiagnosticsStreamPublisher(), 0, DEFAULT_PUBLISH_PERIOD.getSeconds(),
+        TimeUnit.SECONDS);
+  }
+
+  public void stop() throws InterruptedException {
+    scheduler.shutdown();
+
+    // Allow any scheduled publishes to finish, and block for termination (for max 60 seconds)
+    scheduler.awaitTermination(60, TimeUnit.SECONDS);
```

Comment:
Shouldn't block for more than task shutdown ms, ideally much less. Should call shutdownNow() if timeout expires.